warning.rb

```diff
@@ -44,10 +44,16 @@ module Kernel
   # following categories:
   #
   # :deprecated :: Used for warning for deprecated functionality that may
-  #                be removed in the future.
+  #                  be removed in the future.
   # :experimental :: Used for experimental features that may change in
   #                  future releases.
+  # :performance  :: Used for warning about APIs or pattern that have
+  #                  negative performance impact
   def warn(*msgs, uplevel: nil, category: nil)
-    Primitive.rb_warn_m(msgs, uplevel, category)
+    if category.nil?
+      Primitive.rb_warn_m(msgs, uplevel, nil)
+    elsif Warning[Primitive.cexpr!("rb_to_symbol_type(category)")]
```

Comment:
> I believe `Kernel#warn` should behave in the same way than internal `rb_warning_`* APIs

These APIs accept `Symbol` only, not object with `to_sym`, I think.

```suggestion
    elsif Warning[category]
```

Reply:
I thought so as well, but @jeremyevans added a rubyspec to allow `warn msg, category: "deprecated"` in https://github.com/ruby/ruby/pull/3508

Reply:
I see.
The APIs do not accept other than `Symbol`s but `Kernel#warn` method coerces `category` to a `Symbol`.

Another small topic, to get rid of the coercing it twice:

```suggestion
    elsif Warning[category = Primitive.cexpr!("rb_to_symbol_type(category)")]
```